File: src/main/java/jb/data/Dataset.java
package jb.data;

import jb.config.Opts;
import jb.data.clfobj.ClfObject;
import jb.data.clfobj.ClfObjectOnceSorted;
import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.ArrayList;
import java.util.List;

import static java.util.Arrays.sort;
import static jb.util.ModelUtils.getIndexOfSubspace;

@Data
@AllArgsConstructor
public class Dataset {

    private List<Problem> problems;
    private double minX;
    private double maxX;

    public ValidatingTestingTuple getValidatingTestingTuple(Opts opts) {

        Problem basePrblem = problems.get(opts.getPermutation()[0]);
        List<Problem> validatingProblems = new ArrayList<>();
        int[] countOfSubspaceProblem = new int[opts.getNumberOfSpaceParts()];
        ClfObject[] clfObjects = new ClfObjectOnceSorted[basePrblem.getY().length];
        for (int i = 0; i < clfObjects.length; i++) {
            countOfSubspaceProblem[getIndexOfSubspace(opts.getNumberOfSpaceParts(), basePrblem.getX()[i][0], this.minX, this.maxX)]++;
            clfObjects[i] = new ClfObjectOnceSorted(basePrblem.getX()[i], basePrblem.getY()[i]);
        }
        sort(clfObjects);
        int counter = 0;
        for (int i = 0; i < countOfSubspaceProblem.length; i++) {
            double[][] x = new double[countOfSubspaceProblem[i]][basePrblem.getX()[0].length];
            int[] y = new int[countOfSubspaceProblem[i]];
            for (int j = 0; j < countOfSubspaceProblem[i]; j++) {
                x[j] = clfObjects[counter].getX();
                y[j] = clfObjects[counter].getY();
                counter++;
            }
            validatingProblems.add(new Problem(x, y));
        }
        return new ValidatingTestingTuple(validatingProblems, problems.get(opts.getPermutation()[1]), this.minX, this.maxX);
    }

}

File: src/main/java/jb/data/Dataset.kt
package jb.data


import jb.config.Opts
import jb.data.clfobj.ClfObjectOnceSorted
import jb.util.ModelUtils.getIndexOfSubspace

data class Dataset(var problems: List<Problem>, var minX: Double, var maxX: Double) {

    fun getValidatingTestingTuple(opts: Opts): ValidatingTestingTuple {

        val (x1, y1) = this.problems[opts.permutation[0]]
        val validatingProblems = ArrayList<Problem>()
        val countOfSubspaceProblem = IntArray(opts.numberOfSpaceParts)
        val clfObjects = Array(y1.size) {_ -> ClfObjectOnceSorted(DoubleArray(x1[0].size), 0)}
        for (i in clfObjects.indices) {
            countOfSubspaceProblem[getIndexOfSubspace(opts.numberOfSpaceParts, x1[i][0], this.minX, this.maxX)]++
            clfObjects[i] = ClfObjectOnceSorted(x1[i], y1[i])
        }
        clfObjects.sort()
        var counter = 0
        for (i in countOfSubspaceProblem.indices) {
            val x = Array(countOfSubspaceProblem[i]) { DoubleArray(x1[0].size) }
            val y = IntArray(countOfSubspaceProblem[i])
            for (j in 0 until countOfSubspaceProblem[i]) {
                x[j] = clfObjects[counter].x
                y[j] = clfObjects[counter].y
                counter++
            }
            validatingProblems.add(Problem(x, y))
        }
        return ValidatingTestingTuple(validatingProblems, this.problems[opts.permutation[1]], this.minX, this.maxX)
    }

}
